### ISSUE TITLE
Security vulnerabilities in Smallstep

### ISSUE THREAD

**Issue by AGWA** [smallstep/certificates#435]
I did a partial security review of Smallstep and found a couple vulnerabilities and some other things you may want to look at. My findings are here: https://www.agwa.name/blog/post/security_vulnerabilities_in_smallstep

**Comment by maraino** [comment]
Thanks @AGWA, we're aware of the weakness of the AWS Provisioner. Step CA is based on tokens, and AWS identity documents are not provided as tokens as for example GCP does, so we had to embed those documents in a token.

The first unsafe deserialization is used to fetch the document signature that is used in the HMAC.
```go
var unsafeClaims awsPayload
if err := jwt.UnsafeClaimsWithoutVerification(&unsafeClaims); err != nil {
    return nil, errs.Wrap(http.StatusUnauthorized, err, "aws.authorizeToken; error unmarshaling claims")
}
```

Then we validate the token with that signature:
```go 
var payload awsPayload
if err := jwt.Claims(unsafeClaims.Amazon.Signature, &payload); err != nil {
    return nil, errs.Wrap(http.StatusUnauthorized, err, "aws.authorizeToken; error verifying claims")
}
```

At this moment everything is totally insecure as you can generate the token with any document you want and any signature. 

But you're not taking into account the actual validation, when we validate the identity document with the same signature.
```go
if err := p.checkSignature(payload.Amazon.Document, payload.Amazon.Signature); err != nil {
    return nil, errs.Wrap(http.StatusUnauthorized, err, "aws.authorizeToken; invalid aws token signature")
}
```

To pass this validation you need to be in possession of AWS RSA private keys used to sign these documents.

**Comment by AGWA** [comment]
Hi @maraino I agree that the identity document itself is validated as being signed by AWS and I haven't said anything to the contrary.  What is ineffective is the attempt to limit the token lifetime to 5 minutes.  I also find this statement very confusing:

> Then we validate the token with that signature:

What exactly is being validated by that code?

**Comment by maraino** [comment]
> What exactly is being validated by that code?

Just that our software generates the token, it doesn't provide additional security as the real validation is the one with the document. 

The token lifetime is irrelevant, in my opinion, as you can generate a token yourself with any data you want. 

By default, you can only use the first token generated by the instance, the second token will be blocked, the id used is generated based on a verified identity document. 

It is also possible to limit the "validity" of the token to the instance, as you can restrict the token usage window. 

Unfortunately, AWS doesn't provide a better way to use identity documents. We would be able to use shared secrets to sign the token, but this is like using a JWK provisioner, and it will be harder to automate. 

Regarding the JSON injection, this is something to take into account. I've detected the problem in one default X509 template, curiously the one used by IID provisioners. An easy way to fix this is by escaping the quotes using the following template code:
```
"subject": {"commonName":{{ toJson .Insecure.CR.Subject.CommonName }}},
```

**Comment by maraino** [comment]
@AGWA we're open to recommendations of improvements of the AWS provisioner, if you have ideas, share them with us.

**Comment by mmalone** [comment]
Hi @AGWA!

First, thank you for reviewing our code and providing feedback. This level of review is super helpful, and I really appreciate it. I've had a pretty busy morning (dealing with COVID-related childcare issues and it snowed a bunch last night, so just finished digging out the driveway) so I haven't carefully read your post yet. It's literally next up on my todo list though, and I'll try to respond more thoroughly once I've done so.

Regarding your immediate question: for the AWS IID provisioner, _nothing_ is verified by the JWT. The JWT contents are completely insecure. It's all "client-provided data". The fact that JWTs are used at all here was an expediency: as @maraino mentioned, the other clouds send JWTs. AWS uses a proprietary scheme. It was just easier to wrap the proprietary AWS IID in a JWT on the client side and keep the JWT-verification code paths. It is confusing and dumb and we should change this detail at some point, but I don't think there's a security issue here.

Again, I haven't read your post yet, but I skimmed it and I think you've got some other grievances against the AWS IID provisioner. There are several security relevant details to AWS IIDs that are unfortunate, but aren't something we're able to control. We've tried to document these limitations, but we did just rev our docs and it's quite possible that there's room for improvement.

I'm gonna go read your post now!

**Comment by AGWA** [comment]
@maraino I also see unescaped values in `sshutil/templates.go`.  Updating the templates to use `toJson` and emphasizing the need to use `toJson` in your documentation is a good short-term fix, but I must reiterate that your users are going to forget to escape in their own templates, so the proper fix is a context-aware templater for JSON like template/html.

If nothing is being verified by the JWT, I suggest that you remove all the code that tries to verify it - it's just noise and it gives the impression you are trying to uphold security properties.

> By default, you can only use the first token generated by the instance, the second token will be blocked, the id used is generated based on a verified identity document.

As I explained, the server is currently trusting the client to generate a unique ID.  This logic needs to be moved to the server side to be effective.

**Comment by AGWA** [comment]
> I think the right fix is to apply toJson (or another escaping mechanism) to the value of every string-type template variable automatically (number-type variables should be fine as they provide no injection opportunity). We can also provide some sort of unescape function to regain full expressiveness, but with the right default this time. This is the pattern that I've seen in other templating languages, and it seems to work well to prevent accidental misuse. @AGWA do you agree?

Yup, that's the kind of approach I was thinking of.

> >  As I explained, the server is currently trusting the client to generate a unique ID.
>
>Still looking into this.

The problem is this line: https://github.com/smallstep/certificates/blob/1feb4fcb26dc78d70bc1d9e586237a36a8fbea9c/authority/provisioner/aws.go#L287

`payload.ID` is from the JWT, which we've established is untrusted.  The ID should be derived from the contents of the IID, because it's signed by AWS.  (This is why doing unnecessary stuff is a security risk: if JWTs weren't being used, there would be no ID to misuse.)

> Broadly, I wish AWS IIDs were better. It's hard to build anything remotely secure around that mechanism.

I do think the documentation could do a better job explaining the limitations.  It might help to precisely define a threat model so that users know what Smallstep protects against, and so that the design of the AWS provisioner can evaluated against the threat model.

One scenario that really concerns me is a random, less-secured instance in someone's AWS account.  Since it's a random instance, it's not part of the PKI, so it doesn't run Smallstep and has never requested a certificate before.  If an IID leaks from this instance, and DisableCustomSANs hasn't been turned on, an attacker could use that IID get a certificate with literally any identifier - even identifiers which have already been claimed by other instances' certificates.

**Comment by maraino** [comment]
@AGWA: fix on https://github.com/smallstep/certificates/pull/436, I'm trying to see if we should just regenerate it in `GetTokenID()` or just validate it as I'm doing in the current version of the PR

**Comment by mmalone** [comment]
> payload.ID is from the JWT, which we've established is untrusted.

Yup. You're right. @maraino is working on a PR over here: https://github.com/smallstep/certificates/pull/436. We're now debating the (unimportant?) detail of whether the code is clearer if we recompute the value and compare to the user-supplied one or if we ignore the user-supplied value altogether (in which case, why is it even there). Either way, a fix is on its way.

> I do think the documentation could do a better job explaining the limitations.

Yea, you're right. I agree with your critiques of our docs. We're implying something that's simply not true. The description of the `-instance-age` parameter is also confusing. We'll fix both.

Regarding the random unsecured instance: again, 100% agree. When we were designing this mechanism that was one of the corner cases we identified. The max instance age feature was meant to partially mitigate this. We were also _really_ hesitant to make `DisableCustomSANs` default to off. Unfortunately, the mechanism isn't very useful for most people in this configuration.

[Our blog post announcing / describing our IID support](https://smallstep.com/blog/embarrassingly-easy-certificates-on-aws-azure-gcp/) goes into more depth on some of these issues. The model is pretty much "cryptographic perimeter", and that's it. But there's nuance: if you're using trusted code (e.g., config management) to "use up" your IID on each instance quickly after startup (before an attacker has an opportunity to do so) you can probably get to a place where you trust the identity bound to certs. But there are also probably better ways.

Also, we've discussed single-cert-per-identity before. We'll probably do it eventually. But there are operational issues that make it infeasible for pretty much everyone. Perhaps those are addressable. However, we have another idea that I think is better. Medium-term, we're working on better metadata-inventory integrations. A trusted inventory mapping an authenticated instance ID to certificate details will solve a lot of the limitations of the IID mechanism, as currently implemented.

**Comment by dopey** [comment]
Opened up an issue to improve the docs for `instance-age` -- https://github.com/smallstep/docs/issues/12.

**Comment by AGWA** [comment]
> Yup. You're right. @maraino is working on a PR over here: #436. We're now debating the (unimportant?) detail of whether the code is clearer if we recompute the value and compare to the user-supplied one or if we ignore the user-supplied value altogether (in which case, why is it even there). Either way, a fix is on its way.

I think it's more robust to ignore the user-supplied value.  It would also make it easier to remove the JWT down the road.

> Yea, you're right. I agree with your critiques of our docs. We're implying something that's simply not true. The description of the -instance-age parameter is also confusing. We'll fix both.

:+1:

> But there's nuance: if you're using trusted code (e.g., config management) to "use up" your IID on each instance quickly after startup (before an attacker has an opportunity to do so) you can probably get to a place where you trust the identity bound to certs.

Yeah, this isn't a bad way to go, but it should be clear that this is the architecture you're designing for, and any deviation (such as random, unmanaged instances) is risky.

**Comment by mmalone** [comment]
> Yeah, this isn't a bad way to go, but it should be clear that this is the architecture you're designing for, and any deviation (such as random, unmanaged instances) is risky.

Yea, there's probably an interesting meta-story here... we did a first pass at this functionality, and took some care to position it properly when it was unannounced (in the blog post mentioned above). Since then, IIDs haven't been used as much as we expected (ACME is way more popular). I'm sure some people are using it -- it's possible lots of people are and we're just not aware -- but I haven't actually talked to anyone who uses it and we rarely get open issues or support questions about it. So it hasn't gotten much attention and hasn't been a high priority for documentation.

So yea, I appreciate you bringing this to our attention. I think the two places you identified in docs are worth improving. I'm gonna try to come up with some succinct language describing the security characteristics. While I do think the instance age and single-use features could be used to strengthen the security of this mechanism, I'm reluctant to give a hand-wavey description of how that might be done. There are a lot of details you'd need to get right for that to be secure. I think the safe recommendation is to treat this as a "cryptographic perimeter for trusted infrastructure" and not promise too much beyond that. The _additional_ subtlety is that if you disable custom SANs, then this mechanism _is_ pretty secure! So telling the whole story in an accessible way is tricky.

**Comment by mmalone** [comment]
Opened another issue to improve documentation of IID security characteristics at https://github.com/smallstep/docs/issues/13

**Comment by maraino** [comment]
Both JSON injection and AWS token reuse issues are fixed in [v0.15.6](https://github.com/smallstep/certificates/releases/tag/v0.15.6)
